src/librustc_mir/interpret/cast.rs

```diff
@@ -37,37 +37,34 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
         kind: CastKind,
         dest: PlaceTy<'tcx, M::PointerTag>,
     ) -> EvalResult<'tcx> {
-        let src_layout = src.layout;
         use rustc::mir::CastKind::*;
         match kind {
             Unsize => {
                 self.unsize_into(src, dest)?;
             }
 
             Misc => {
+                let src_layout = src.layout;
                 let src = self.read_value(src)?;
 
-                if M::ENABLE_PTR_TRACKING_HOOKS &&
-                    src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr()
-                {
+                let src = if M::ENABLE_PTR_TRACKING_HOOKS && src_layout.ty.is_region_ptr() {
+                    assert!(dest.layout.ty.is_unsafe_ptr());
```

Comment:
add a comment mentioning that the only `Misc` casts that start out with a reference must be to a raw pointer. Clear in hindsight, but took me a minute to figure out why this assert is dead code barring bugs (which is exactly what it's trying to catch)